src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,392 @@
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new block and reached the consensus tip 
+                // or the max batch size is reached or the node is shutting down.  
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    BlockPair item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    batch.Add(item);
+
+                    this.currentBatchSizeBytes += item.Block.GetSerializedSize();
+
+                    saveBatch = saveBatch || (item.ChainedHeader == this.chain.Tip) || (this.currentBatchSizeBytes >= BatchThresholdSizeBytes);
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (batch.Count != 0)
+                    {
+                        await this.SaveBatchAsync(batch).ConfigureAwait(false);
+
+                        batch.Clear();
+                        this.currentBatchSizeBytes = 0;
+                    }
+
+                    timerTask = null;
+                }
+                else
+                {
+                    // Start timer if it is not started already.
+                    timerTask = timerTask ?? Task.Delay(BatchMaxSaveIntervalSeconds * 1000, this.nodeLifetime.ApplicationStopping);
+                }
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>Checks if repository contains reorged blocks and deletes them; saves batch on top.</summary>
+        /// <param name="batch">List of batched blocks. Cannot be empty.</param>
+        private async Task SaveBatchAsync(List<BlockPair> batch)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(batch), nameof(batch.Count), batch.Count);
+
+            List<BlockPair> clearedBatch = this.GetBatchWithoutReorgedBlocks(batch);
+
+            ChainedHeader expectedStoreTip = clearedBatch.First().ChainedHeader.Previous;
+
+            // Check if block repository contains reorged blocks. If it does - delete them.
+            if (expectedStoreTip.HashBlock != this.storeTip.HashBlock)
```

Comment:
>This code could potentiall happen before this.GetBatchWithoutReorgedBlocks(batch); no?

It can. But the action upon that condition require first block from cleared batch to be passed inside `RemoveReorgedBlocksFromStoreAsync` so it's not possible to change the order and first remove reorged blocks and after that clean the batch

>Also are we sure the last item in the batch (or first) is the main chain, how can we make that assumption?

First might not be. But in cleared batch- it always is. 

Last item inside the batch is always on the main chain at the moment of saving the batch. It might be reorged by the next batch though, it's ok. 

The item is on the best chain because when consensus reorgs it signals all blocks from the fork point in their right order and those blocks get to the batch

Reply:
`Last item inside the batch is always on the main chain` Ahh so we are using the last item and assuming thats always on the current chain, we delete anything that is not a previous block of the last item.

So store completely relays on blocks sent from consensus to determine the tip?

Reply:
>So store completely relays on blocks sent from consensus to determine the tip?

Correct. 

And in case at the moment of a save it's wrong- we will have replacement blocks in the next batch anyway since consensus loop will provide all of them from the forkpoint.